### PR TITLE
bpf: Add 'host not ready' drop reason if host policy cannot be enforced

### PR DESCRIPTION
**Pull request by ti-mo**
Repurpose unused drop reasons since https://github.com/cilium/cilium/commit/61fb508f2c7320e6cf29f0024f678e5ea22e5d15
("bpf: Rename unused drop defines to DROP_UNUSED*") for signaling the host
endpoint's policy program was attempted to be executed before it was loaded.

bpf_lxc.c contains multiple tail calls into POLICY_CALL_MAP at the HOST_EP_ID
slot. The program in this slot is provided by bpf_host.c. During first agent
startup, there are often multiple Pods pending creation due to no CNI being
available. As soon as the agent's local API becomes available, these outstanding
CNI requests have a chance to be accepted by the API handler.

If one such request is serviced before the host datapath controller has a chance
to grab the compilation lock, the endpoint program will compile and attach first.
If the host firewall is enabled, and this new workload endpoint sends a packet
to the host, the host's ingress policy needs to be enforced. However, because
bpf_host hasn't been loaded yet, this policy program is not yet present in
POLICY_CALL_MAP, resulting in a missed tail call.

One potential solution to this problem would be making sure the host datapath
always attaches before workload endpoints do. There's one problem with this
solution: clustermesh requires data from other clusters in order to correctly
populate the local ipcache, and the ipcache currently needs to be populated for
the host endpoint to finish attaching. It obtains this information through
clustermesh-apiserver, typically deployed onto the local node as a regular Pod.
This means workload endpoints must be able to deploy before the host endpoint.

As a stop gap, tolerate these kinds of drops and assign them a specific meaning,
without letting them spill over into the generic 'missed tail call' bucket. To
stabilize end-to-end tests, we're aiming to enforce zero dropped tail calls in
all CI scenarios, since it leads to packets that mysteriously go missing,
introducing chaos that's impossible to troubleshoot.

```release-note
Add specific drop reason for missing tail calls if the host datapath is not ready yet
```

### CODE REVIEW

**Review by aanm — COMMENTED** [review]
Don't we also need to wait for the hostdatapath to be initialized during restoration of endpoints?

**Comment by ti-mo** [comment]
> Don't we also need to wait for the hostdatapath to be initialized during restoration of endpoints?

Technically, yes. If the agent is restarted, the assumption is for the old host datapath to still be present, but if the agent crashed or was restarted during the initial bringup, this may still result in missed tail calls under relatively normal operations. I'll try to include this.

Looks like there's quite a few tests failing, so this needs more work.

File: daemon/cmd/state.go [hunk 1]
```diff
@@ -318,39 +318,30 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
...
-				} else {
-					epRegenerated <- true
-				}
-				break
-			}
+	for _, ep := range state.restored {
+		// Cap the timeout used to wait for remote cluster synchronization
+		// to avoid blocking the agent startup, as this regeneration is
+		// performed synchronously.
+		endpointsRegenerator.CapTimeoutForSynchronousRegeneration()
```
Comment (tommyp1ckles) [inline daemon/cmd/state.go]: The function comment for `CapTimeoutForSynchronousRegeneration` still says this is only used when ipsec is enabled.

File: daemon/cmd/state.go [hunk 2]
```diff
@@ -318,39 +318,30 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
...
+			epRegenerated <- false
+		} else {
+			epRegenerated <- true
 		}
+		break
 	}
 
 	for _, ep := range state.restored {
-		if ep.IsHost() && option.Config.EnableIPSec {
+		if ep.IsHost() {
```
Comment (tommyp1ckles) [inline daemon/cmd/state.go]: Why do we not skip regenerating here again if the sync one happened above?

Comment (ti-mo) [reply daemon/cmd/state.go]: Sorry, I don't understand the question. The first loop handles the host endpoint, this loop handles all the others.

File: daemon/cmd/state.go (hunk 1 above)
Comment (ti-mo) [reply daemon/cmd/state.go]: Hmm, that might explain the clustermesh failures. I'll lift it out of the loop,

File: daemon/cmd/state.go (hunk 2 above)
Comment (ti-mo) [reply daemon/cmd/state.go]: Removed this code for now, it may be the cause of clustermesh test failures.

**Review by dylandreimerink — APPROVED** [review]

File: daemon/cmd/endpoint.go [hunk 3]
```diff
@@ -569,6 +569,12 @@ func putEndpointIDHandler(d *Daemon, params PutEndpointIDParams) (resp middlewar
 	}
 	defer r.Done()
 
+	// Wait for the host datapath to be initialized before starting to manage
+	// endpoint programs. bpf_lxc makes the assumption that bpf_host is loaded
+	// and its policy program(s) inserted. Can't live in createEndpoint as that
+	// function is called from tests that don't set up a host datapath.
+	<-d.Datapath().Loader().HostDatapathInitialized()
```
Comment (tommyp1ckles) [inline daemon/cmd/endpoint.go]: Makes sense to me,  blocking inside the api handler does seem a bit strange to me however.

One possible improvement on this could be to `select {}` on both this and the request context in case the request is cancelled prior to this completing.

```go
select {
    case <-d.Datapath().Loader().HostDatapathInitialized():
    case <-d.apiLimiterSet.Wait(params.HTTPRequest.Context().Done():
        // ...
}
```

Comment (ti-mo) [reply daemon/cmd/endpoint.go]: The code in question has been removed.

File: api/v1/flow/flow.proto [hunk 4]
```diff
@@ -335,7 +335,7 @@ enum DropReason {
     // non-drop reasons
     DROP_REASON_UNKNOWN = 0;
     // drop reasons
-    INVALID_SOURCE_MAC = 130;
+    HOST_NOT_READY = 130;
```
Comment (rolinh) [inline api/v1/flow/flow.proto]: Don't reuse a field number as it messes up deserialization. A client with an older definition of the API would report `INVALID_SOURCE_MAC` and not `HOST_NOT_READY`.
Best practice here if this field is unused is to  mark it as deprecated like below and pick the next available field number for `HOST_NOT_READY`.
```suggestion
    // Deprecated: this drop reason is not emitted since Cilium version x.y.z
    HOST_NOT_READY = 130 [deprecated = true];
```

**Review by rolinh — CHANGES_REQUESTED** [review]

**Comment by pchaigno** [comment]
FYI, I believe you will need to rebase to catch https://github.com/cilium/cilium/pull/29942, because that CLI release introduced a new test that fails on unexpected packet drop reasons. Then either you allowlist the new drop reason by default in the CLI repository (which we probably want to do either way) or via the `--expected-drop-reason="+Host not ready"` flag here.

**Review by rolinh — APPROVED** [review]
API changes lgtm.

**Comment by ti-mo** [comment]
> FYI, I believe you will need to rebase to catch #29942, because that CLI release introduced a new test that fails on unexpected packet drop reasons. Then either you allowlist the new drop reason by default in the CLI repository (which we probably want to do either way) or via the `--expected-drop-reason="+Host datapath not ready"` flag here.

Thanks for the heads up. What are the workflows I need to add that expected drop reason to? Assuming the new default is added to the CLI this week, it won't be released (and bumped in c/c, and backported) for a while.

**Comment by pchaigno** [comment]
> What are the workflows I need to add that expected drop reason to?

Only those using the connectivity tests, of which I think only `conformance-e2e.yaml` has the host firewall enabled. If we miss any, I guess we'll know once the rebased tests run.

File: bpf/bpf_lxc.c [hunk 5]
```diff
@@ -1053,11 +1053,13 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
 	/* If the destination is the local host and per-endpoint routes are
 	 * enabled, jump to the bpf_host program to enforce ingress host policies.
+	 * Note that bpf_lxc can be loaded before bpf_host, so bpf_host's policy
+	 * program may not yet be present at this time.
 	 */
 	if (*dst_sec_identity == HOST_ID) {
 		ctx_store_meta(ctx, CB_FROM_HOST, 0);
 		tail_call_static(ctx, &POLICY_CALL_MAP, HOST_EP_ID);
-		return DROP_MISSED_TAIL_CALL;
+		return DROP_HOST_NOT_READY;
```
Comment (tommyp1ckles) [inline bpf/bpf_lxc.c:1065]: Will this case, and the `send_drop_notify` case below always be as a result of the host policy program not being in place?

File: api/v1/flow/flow.proto [hunk 6]
```diff
@@ -335,15 +335,15 @@ enum DropReason {
     // non-drop reasons
     DROP_REASON_UNKNOWN = 0;
     // drop reasons
-    INVALID_SOURCE_MAC = 130;
-    INVALID_DESTINATION_MAC = 131;
+    INVALID_SOURCE_MAC = 130 [deprecated = true];
```
Comment (tommyp1ckles) [inline api/v1/flow/flow.proto:338]: Are these deprecations a consequence of the new drop reason?  If not it might be worth putting this into a separate commit to separate-out & explain the change.

Comment (ti-mo) [reply api/v1/flow/flow.proto:338]: I initially lobbed it in to avoid regenerating the proto twice, easing backports, but with the IGMP drop reasons  added in between we'll need to regenerate when backporting anyway.

File: bpf/bpf_lxc.c (hunk 5 above)
Comment (ti-mo) [reply bpf/bpf_lxc.c:1065]: Yes, if calling into POLICY_CALL_MAP at slot HOST_EP_ID fails, the host program has not been loaded.

**Review by brlbil — APPROVED** [review]
CI part LGTM.

**Review by nathanjsweet — APPROVED** [review]
